### PR TITLE
[MoE] Using routing map instead of histc to count number_tokens_per_expert

### PR DESCRIPTION
**Pull request by acisseJZhong**
## What changed
 1. Router.forward() now builds and returns an explicit boolean routing_map_BLE with shape (B, L, E), using scatter_() (using a local map to wrap around scatter_ becuase scatter_ doesn't support DTensor yet) from topk_expert_ids_BLK.
 2. Per-expert token counts are computed from `routing_map.sum(...)` instead of `torch.histc(...)`.

## Test plan

Loss comparison: The loss is bit-wise identical across 100 stesps
```
  NCCL_NVLS_ENABLE=0 \
  CUDA_VISIBLE_DEVICES=4,5,6,7 \
  python scripts/loss_compare.py . main \
    --baseline-module='llama4' \
    --baseline-config='llama4_debugmodel' \
    --baseline-options='--parallelism.pipeline_parallel_degree 1 --parallelism.data_parallel_shard_degree 2
  --parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 2' \
    --baseline-ngpus=4 \
    --test-module='llama4' \
    --test-config='llama4_debugmodel' \
    --test-options='--parallelism.pipeline_parallel_degree 1 --parallelism.data_parallel_shard_degree 2
  --parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 2' \
    --test-ngpus=4 \
    --no-seed-checkpoint \
    --output-folder=agent_space/loss_compare_current_vs_main_dp2tp2ep2
```

### CODE REVIEW

File: torchtitan/models/common/moe.py [hunk 1]
```diff
@@ -92,6 +109,7 @@ def forward(
         x_BLD: torch.Tensor,
         topk_scores_BLK: torch.Tensor,
         topk_expert_ids_BLK: torch.Tensor,
+        routing_map_BLE: torch.Tensor,
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: In fact, I think you can just pass in `num_tokens_per_expert_E`

File: torchtitan/models/common/moe.py [hunk 2]
```diff
@@ -274,15 +297,36 @@ def forward(
...
+                out_placements=(scores_BLE.placements,),
+                device_mesh=scores_BLE.device_mesh,
+            )
+        else:
+            generate_routing_map = _generate_routing_map_ble
+
+        routing_map_BLE = generate_routing_map(  # pyrefly: ignore [bad-argument-count]
+            scores_BLE, topk_expert_ids_BLK
         )
+        num_tokens_per_expert_E = routing_map_BLE.sum(dim=(0, 1))
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: Let's
- return `scores_BLE` (which will be needed by https://github.com/pytorch/torchtitan/pull/3000/changes#diff-9c3e17fd22b0e249ef06806990a6a33b74d45f9ce009100d3a040e27ef074a3eR306 anyways), not `routing_map_BLE` or `num_tokens_per_expert_E`
- compute them in MoE fwd, not in router

File: torchtitan/models/common/moe.py [hunk 3]
```diff
@@ -378,7 +423,9 @@ def forward(self, x_BLD: torch.Tensor) -> torch.Tensor:
         with torch.no_grad():
             self.tokens_per_expert_E.add_(num_tokens_per_expert_E)
 
-        out_TD = self.experts(x_BLD, topk_scores_BLK, topk_expert_ids_BLK)
+        out_TD = self.experts(
+            x_BLD, topk_scores_BLK, topk_expert_ids_BLK, routing_map_BLE
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: let's pass `num_tokens_per_expert_E` directly.

File: torchtitan/models/common/moe_sharding.py [hunk 4]
```diff
@@ -250,6 +250,7 @@ def set_moe_sharding_config(
             "x_BLD": experts_in_layout,
             "topk_scores_BLK": experts_in_layout,
             "topk_expert_ids_BLK": experts_in_layout,
+            "routing_map_BLE": experts_in_layout,
```
Comment (tianyu-l) [inline torchtitan/models/common/moe_sharding.py]: Let's say we pass `num_tokens_per_expert_E` here. It will be used in `offsets` of `grouped_mm`. I think this field won't have gradient. Thus the gradient placement will have no effect. Let's specify None here and see if the code runs and numerics still match.

File: torchtitan/models/common/moe.py [hunk 5]
```diff
@@ -274,15 +297,36 @@ def forward(
...
-            max=self.num_experts,
+        # Build routing map with scatter. scatter_ does not support mixed
+        # local Tensor / DTensor arguments, so run the scatter on local tensors
+        # under local_map when router outputs are DTensors.
+        if isinstance(topk_expert_ids_BLK, DTensor):
+            assert isinstance(
+                scores_BLE, DTensor
+            ), "scores_BLE and topk_expert_ids_BLK should both be DTensors"
+            generate_routing_map = local_map(
+                _generate_routing_map_ble,
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: maybe define this function inline here

File: torchtitan/models/common/moe.py [hunk 6]
```diff
@@ -274,15 +297,36 @@ def forward(
...
 
-        # group tokens together by expert indices from 0 to num_experts and pass that to experts forward
-        num_tokens_per_expert_E = torch.histc(
-            topk_expert_ids_BLK.view(-1),
-            bins=self.num_experts,
-            min=0,
-            max=self.num_experts,
+        # Build routing map with scatter. scatter_ does not support mixed
+        # local Tensor / DTensor arguments, so run the scatter on local tensors
+        # under local_map when router outputs are DTensors.
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: leave a TODO to support sharding prop for scatter

File: torchtitan/models/common/moe_sharding.py (hunk 4 above)
Comment (wwwjn) [reply torchtitan/models/common/moe_sharding.py]: I see, the `num_tokens_per_expert_E` shouldn't have gradient because it's just routing information.

File: torchtitan/models/common/moe.py [hunk 7]
```diff
@@ -109,7 +110,9 @@ def forward(
             routed_input_RD,
             num_tokens_per_expert_E,
             metadata,
-        ) = self.token_dispatcher.dispatch(x_TD, topk_scores_TK, topk_expert_ids_TK)
+        ) = self.token_dispatcher.dispatch(
+            x_TD, topk_scores_TK, topk_expert_ids_TK, num_tokens_per_expert_E
+        )
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:123]: why sending `num_tokens_per_expert_E` to dispatch() at all?

File: torchtitan/models/common/token_dispatcher.py [hunk 8]
```diff
@@ -595,7 +604,9 @@ def dispatch(
         x_TD: torch.Tensor,
         topk_scores_TK: torch.Tensor,
         topk_expert_ids_TK: torch.Tensor,
+        num_tokens_per_expert_E: torch.Tensor,
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: can we not pass it in?

File: torchtitan/models/common/moe.py (hunk 7 above)
Comment (wwwjn) [reply torchtitan/models/common/moe.py:123]: Oh i see, the issue. I was trying to remove `histc` usage in LocalTokenDispatcher by passing the `num_tokens_per_expert_E` as well, but acutally is not necessary. We can still use `histc` in LocalTokenDispatcher. Updated

Comment (tianyu-l) [reply torchtitan/models/common/moe.py:123]: No. Where is num_tokens_per_expert_E / histc used in dispatcher?

File: torchtitan/models/common/token_dispatcher.py [hunk 9]
```diff
@@ -88,13 +88,12 @@ def _local_reorder(
...
         """
-        # group tokens together by expert indices from 0 to num_experts and pass that to experts forward
         num_tokens_per_expert_E = torch.histc(
-            topk_expert_ids_TK.view(-1),
+            topk_expert_ids_TK.float(),
             bins=self.num_experts,
             min=0,
             max=self.num_experts,
-        )
+        ).to(torch.int64)
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: maybe I'm missing something -- could you explain why we need to compute `num_tokens_per_expert_E` in dispatcher and return it AT ALL?

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/models/common/token_dispatcher.py (hunk 9 above)
Comment (wwwjn) [reply torchtitan/models/common/token_dispatcher.py]: mb, I've already moved the calculation in `MoE` level and we don't need to calculate in dispatcher level. I should clean up this part cleanly

But we still need to pass `num_tokens_per_expert_E` into dispatch() API and return from it. This is because for `num_tokens_per_expert_E` in DeepEPDispatcher path actually comes from DeepEP kernel's return value. If DeepEP is applied, we still need to return. To be more clear, the current implementation is: 
1. For AllToAllTokenDispatcher we pass `num_tokens_per_expert_E` into dispatch() call and return it as is.
2. For DeepEPTokenDispatcher, we use the DeepEP returned `num_tokens_per_expert_E`, instead of what we passed into dispatch().

An alternative is don't pass `num_tokens_per_expert_E` into dispatch() at all,  just drop the`num_tokens_per_expert_E`  from DeepEP and use the value calculated in `MoE`.  Can we take the alternative way? Is it possible that DeepEP returns a different value than what we calculated from routing_map?

File: torchtitan/models/common/moe.py [hunk 10]
```diff
@@ -363,13 +363,50 @@ def forward(self, x_BLD: torch.Tensor) -> torch.Tensor:
...
-            num_tokens_per_expert_E,
+            scores_BLE,
         ) = self.router(x_BLD, self.expert_bias_E)
 
+        # Build routing map with scatter. scatter_ does not support mixed
+        # local Tensor / DTensor arguments, so run the scatter on local tensors
+        # under local_map when router outputs are DTensors.
+        # TODO: Remove this local_map workaround once DTensor sharding
+        # propagation supports scatter with mixed Tensor / DTensor arguments.
+        def generate_routing_map_ble(
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]:
```suggestion
        def _generate_routing_map(
```

File: torchtitan/models/common/token_dispatcher.py
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: change `tokens_per_expert_E` variable name
same for DeepEP

Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py:326]: add a TODO that we might want to use this as buffer to compute `expert_bias_E`, but HybridEP token dropping is not clear.

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/common/moe.py [hunk 11]
```diff
@@ -53,45 +54,48 @@ def __init__(self, config: Config):
     def _experts_forward(
         self,
         x_RD: torch.Tensor,
-        num_tokens_per_expert_E: torch.Tensor,
+        num_global_tokens_per_local_expert_e: torch.Tensor,
```
Comment (wwwjn) [inline torchtitan/models/common/moe.py]: @tianyu-l  Can I get a final review of the rename in `_experts_forward` function? Basically I renamed `E` -> `e`

Comment (tianyu-l) [reply torchtitan/models/common/moe.py]: Good point. I discussed with @acisseJZhong before.

In grouped experts:
- if EP is used, then this notation is correct (e is sharded version of E)
- if TP is used, then we should do `w1_EfD` instead because TP shards on different things
- maybe we should use x_rD instead of x_RD depending on batch/seq dim sharding

Given all these complications, let's just use capital letters (such as `E`) and `num_tokens_per_expert_E`.

Comment (tianyu-l) [reply torchtitan/models/common/moe.py]: good to add a NOTE on notation

File: torchtitan/models/common/moe_sharding.py [hunk 12]
```diff
@@ -258,6 +261,9 @@ def set_moe_sharding_config(
                 experts_in_grad_layout,
                 experts_in_grad_layout,
                 experts_in_grad_layout,
+                # num_local_tokens_per_expert_E is routing metadata, but it is
+                # still a DTensor input to local_map and must have placements.
+                _tokens_per_expert_placement(enable_ep=enable_ep),
```
Comment (wwwjn) [inline torchtitan/models/common/moe_sharding.py:266]: I have to put a placement here, because local_map doesn't take None as a input, cc @tianyu-l

Comment (tianyu-l) [reply torchtitan/models/common/moe_sharding.py:266]: it takes?
https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_func_map.py#L20-L29

File: torchtitan/models/common/moe.py (hunk 11 above)
Comment (wwwjn) [reply torchtitan/models/common/moe.py]: Sounds good, 

> let's just use capital letters (such as E) and `num_tokens_per_expert_E`.

I updated `num_tokens_per_expert_E` to be the parameter name of _expert_forwarrd(), but in the callsite, I still pass `num_tokens_per_expert_E=num_global_tokens_per_local_expert_e` to distinguish: 
```
(
    routed_input_RD,
    num_global_tokens_per_local_expert_e,
    metadata,
) = self.token_dispatcher.dispatch(
    x_TD, topk_scores_TK, topk_expert_ids_TK, num_local_tokens_per_expert_E
)
routed_output_RD = self._experts_forward(
    routed_input_RD, num_global_tokens_per_local_expert_e
)
```

Comment (tianyu-l) [reply torchtitan/models/common/moe.py]: that sounds good
